cmd/signature-v2.go

```diff
@@ -191,14 +192,33 @@ func doesSignV2Match(r *http.Request) APIErrorCode {
 		return apiError
 	}
 
-	// url.RawPath will be valid if path has any encoded characters, if not it will
-	// be empty - in which case we need to consider url.Path (bug in net/http?)
+	// Encode path:
+	//   url.RawPath will be valid if path has any encoded characters, if not it will
+	//   be empty - in which case we need to consider url.Path (bug in net/http?)
 	encodedResource := r.URL.RawPath
-	encodedQuery := r.URL.RawQuery
 	if encodedResource == "" {
 		splits := strings.Split(r.URL.Path, "?")
 		if len(splits) > 0 {
-			encodedResource = splits[0]
+			encodedResource = getURLEncodedName(splits[0])
+		}
+	}
+
+	// Encode query strings
+	encodedQuery := ""
```

Comment:
> why can't we get like r.URL.Query().Encode() instead of looping and creating our own encodedQuery?

actually, this does more than encoding like sorting and removing "=" in case of empty values for some parameters.. but also I found this is done in lower layer. so I simplified as you suggested